In [3]:
# imports
import pandas as pd
import seaborn as sns

In [4]:
# load in Ahana's final merged dataset
total_data = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/Totaldata_Jun_21.csv"
merged_df = pd.read_csv(total_data, encoding = "ISO-8859-1")

In [5]:
# checkout merged_df
print("Shape of full dataset: {}".format(merged_df.shape))
merged_df.sample(10)

Shape of full dataset: (18855, 64)


Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY  \
9715         9715  200013307  Total   13307  2000     13     307   
1002         1002  201212101  Total   12101  2012     12     101   
10434       10434  200025023  Total   25023  2000     25      23   
11377       11377  200040073  Total   40073  2000     40      73   
17454       17454  200837145  Total   37145  2008     37     145   
10940       10940  200031161  Total   31161  2000     31     161   
13484       13484  200422061  Total   22061  2004     22      61   
6556         6556  201655101  Total   55101  2016     55     101   
17966       17966  200847063  Total   47063  2008     47      63   
16933       16933  200828081  Total   28081  2008     28      81   

               STNAME            CTYNAME  AA_FEMALE        ...          \
9715          Georgia     Webster County          0        ...           
1002          Florida       Pasco County       5658        ...           
10434   Massachusetts    Plymouth County       2510        ...           
11377        Oklahoma  Kingfisher County         18        ...           
17454  North Carolina      Person County         59        ...           
10940        Nebraska    Sheridan County          7        ...           
13484       Louisiana     Lincoln Parish        285        ...           
6556        Wisconsin      Racine County       1419        ...           
17966       Tennessee     Hamblen County        232        ...           
16933     Mississippi         Lee County        261        ...           

       NA_MALE%  WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  \
9715   0.000000    0.253975  0.269456     0.497490   0.502510   
1002   0.000451    0.466811  0.441612     0.513899   0.486101   
10434  0.000307    0.469839  0.444432     0.512729   0.487271   
11377  0.000000    0.475883  0.452053     0.512561   0.487439   
17454  0.000000    0.359349  0.342793     0.517151   0.482849   
10940  0.000000    0.450857  0.436793     0.510346   0.489654   
13484  0.000207    0.288339  0.290636     0.513126   0.486874   
6556   0.000308    0.424666  0.416839     0.504848   0.495152   
17966  0.000552    0.475029  0.453500     0.510716   0.489284   
16933  0.000112    0.369385  0.346582     0.521046   0.478954   

       TOT_POP_LESS19%  TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  \
9715          0.281590         0.254393         0.268201          0.195816   
1002          0.233627         0.217313         0.274346          0.274715   
10434         0.291018         0.265253         0.288173          0.155556   
11377         0.299742         0.236291         0.265360          0.198608   
17454         0.264668         0.227675         0.307585          0.200072   
10940         0.279179         0.185742         0.270934          0.264145   
13484         0.297894         0.338478         0.215186          0.148442   
6556          0.261501         0.240047         0.288288          0.210165   
17966         0.259076         0.254351         0.277228          0.209345   
16933         0.296756         0.262658         0.270093          0.170492   

       Unemployment_Rate  
9715                 3.8  
1002                 9.0  
10434                2.7  
11377                2.2  
17454                7.2  
10940                2.9  
13484                6.5  
6556                 4.9  
17966                7.1  
16933                6.9  

[10 rows x 64 columns]

In [6]:
# see all columns in merged dataframe
print(merged_df.columns)

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       '

In [7]:
# see how many nulls there are within dataframe
merged_df.isnull().sum()

Unnamed: 0                 0
ID                         0
AGEGRP                     0
FIPS_x                     0
YEAR                       0
STATE                      0
COUNTY                     0
STNAME                     0
CTYNAME                    0
AA_FEMALE                  0
AA_MALE                    0
BA_FEMALE                  0
BA_MALE                    0
H_FEMALE                   0
H_MALE                     0
IA_FEMALE                  0
IA_MALE                    0
NA_FEMALE                  0
NA_MALE                    0
SUMLEV                     0
TOT_FEMALE                 0
TOT_MALE                   0
TOT_POP                    0
WA_FEMALE                  0
WA_MALE                    0
TOT_POP_LESS19             0
TOT_MALE_LESS19            0
TOT_FEMALE_LESS19          0
TOT_POP_20to39             0
TOT_MALE_20to39            0
                        ... 
TOT_POP_Above60            0
TOT_MALE_Above60           0
TOT_FEMALE_Above60         0
republican_can

### One hot encode "winning party" column to binary feature
* republican = 1
* democrat = 0

In [8]:
winning_party_binary = pd.get_dummies(merged_df["winning party"], drop_first = True)
merged_df["winning_party_binary"] = winning_party_binary

In [9]:
merged_df.sample(10)

Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY      STNAME  \
873           873   20206111  Total    6111  2020      6     111  California   
16016       16016  200816005  Total   16005  2008     16       5       Idaho   
3703         3703  201255049  Total   55049  2012     55      49   Wisconsin   
3518         3518  201251139  Total   51139  2012     51     139    Virginia   
15823       15823  200812083  Total   12083  2008     12      83     Florida   
5558         5558  201639115  Total   39115  2016     39     115        Ohio   
4288         4288  201619083  Total   19083  2016     19      83        Iowa   
18460       18460  200851159  Total   51159  2008     51     159    Virginia   
16490       16490   20082105  Total    2105  2008      2     105      Alaska   
3401         3401  201248493  Total   48493  2012     48     493       Texas   

                         CTYNAME  AA_FEMALE          ...           WA_FEMALE%  \
873               Ventura County      36235          ...             0.423105   
16016             Bannock County        538          ...             0.466283   
3703                 Iowa County         78          ...             0.488879   
3518                 Page County         55          ...             0.490942   
15823              Marion County       2360          ...             0.437190   
5558               Morgan County         12          ...             0.465943   
4288               Hardin County         64          ...             0.485200   
18460            Richmond County         15          ...             0.318799   
16490  Hoonah-Angoon Census Area         11          ...             0.214541   
3401               Wilson County        136          ...             0.479876   

       WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  TOT_POP_20to39%  \
873    0.417894     0.504894   0.495106         0.254674         0.265402   
16016  0.458633     0.503542   0.496458         0.309032         0.300260   
3703   0.491833     0.498670   0.501330         0.267421         0.217659   
3518   0.473573     0.508227   0.491773         0.236475         0.222056   
15823  0.409914     0.517971   0.482029         0.221393         0.199694   
5558   0.464518     0.500814   0.499186         0.240095         0.213569   
4288   0.476372     0.498009   0.501991         0.246264         0.212913   
18460  0.347019     0.434755   0.565245         0.198281         0.267133   
16490  0.263158     0.464318   0.535682         0.239518         0.201606   
3401   0.476810     0.502473   0.497527         0.288130         0.211835   

       TOT_POP_40to59%  TOT_POP_Above60%  Unemployment_Rate  \
873           0.261963          0.217961                NaN   
16016         0.242910          0.147798                4.7   
3703          0.312624          0.202296                6.7   
3518          0.296435          0.245034               10.2   
15823         0.261908          0.317006                8.1   
5558          0.279647          0.266689                8.3   
4288          0.263112          0.277710                3.9   
18460         0.300764          0.233821                5.0   
16490         0.375558          0.183318                NaN   
3401          0.310310          0.189725                5.5   

       winning_party_binary  
873                       0  
16016                     1  
3703                      0  
3518                      1  
15823                     1  
5558                      1  
4288                      1  
18460                     0  
16490                     0  
3401                      1  

[10 rows x 65 columns]

In [10]:
merged_df[["winning party", "winning_party_binary"]].sample(10)

winning party  winning_party_binary
17285      democrat                     0
8471            NaN                     0
4691     republican                     1
6293     republican                     1
15313    republican                     1
223        democrat                     0
14860    republican                     1
6918            NaN                     0
13715    republican                     1
5898     republican                     1

In [11]:
election_years = [2020]

F20_df = merged_df[merged_df["YEAR"].isin(election_years)]
print("2020 data dimensions: {}".format(F20_df.shape))

2020 data dimensions: (3142, 65)


In [97]:
election_years = [2016]

F16_df = merged_df[merged_df["YEAR"].isin(election_years)]
print("2020 data dimensions: {}".format(F16_df.shape))

2020 data dimensions: (3142, 65)


In [12]:
F20_df.sample(10)

Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY       STNAME  \
7971        7971  202031067  Total   31067  2020     31      67     Nebraska   
7723        7723  202028077  Total   28077  2020     28      77  Mississippi   
8499        8499  202041015  Total   41015  2020     41      15       Oregon   
7425        7425  202022057  Total   22057  2020     22      57    Louisiana   
8458        8458  202040087  Total   40087  2020     40      87     Oklahoma   
7911        7911  202030059  Total   30059  2020     30      59      Montana   
7876        7876  202029221  Total   29221  2020     29     221     Missouri   
7207        7207  202020071  Total   20071  2020     20      71       Kansas   
689          689   20201115  Total    1115  2020      1     115      Alabama   
8766        8766  202047109  Total   47109  2020     47     109    Tennessee   

                CTYNAME  AA_FEMALE          ...           WA_FEMALE%  \
7971        Gage County         69          ...             0.483227   
7723    Lawrence County         23          ...             0.334564   
8499       Curry County        120          ...             0.467277   
7425   Lafourche Parish        430          ...             0.408317   
8458     McClain County        142          ...             0.432662   
7911     Meagher County          8          ...             0.476420   
7876  Washington County         40          ...             0.470633   
7207     Greeley County          2          ...             0.507742   
689    St. Clair County        428          ...             0.444300   
8766     McNairy County         51          ...             0.466282   

      WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  TOT_POP_20to39%  \
7971  0.481506     0.500535   0.499465         0.246080         0.211464   
7723  0.327258     0.514733   0.485267         0.264633         0.229225   
8499  0.453338     0.507868   0.492132         0.158462         0.166616   
7425  0.395862     0.509525   0.490475         0.257341         0.267217   
8458  0.426160     0.503714   0.496286         0.278504         0.240990   
7911  0.489818     0.494641   0.505359         0.209003         0.170418   
7876  0.481378     0.485306   0.514694         0.249128         0.235737   
7207  0.468623     0.520782   0.479218         0.282804         0.224124   
689   0.433352     0.503586   0.496414         0.246713         0.249025   
8766  0.448242     0.508671   0.491329         0.238619         0.220308   

      TOT_POP_40to59%  TOT_POP_Above60%  Unemployment_Rate  \
7971         0.258782          0.283674                NaN   
7723         0.254597          0.251546                NaN   
8499         0.232280          0.442642                NaN   
7425         0.258350          0.217092                NaN   
8458         0.262774          0.217732                NaN   
7911         0.221329          0.399250                NaN   
7876         0.276350          0.238784                NaN   
7207         0.213529          0.279544                NaN   
689          0.272996          0.231266                NaN   
8766         0.264827          0.276247                NaN   

      winning_party_binary  
7971                     0  
7723                     0  
8499                     0  
7425                     0  
8458                     0  
7911                     0  
7876                     0  
7207                     0  
689                      0  
8766                     0  

[10 rows x 65 columns]

### Explore if we could drop null values
* merged_df before dropping null values has: 18,855 rows and 66 columns
* merged_df after dropping null values has: 15,437 rows and 66 columns

In [13]:
merged_df2 = merged_df.dropna()
merged_df2.shape

(15569, 65)

In [15]:
merged_df2.sample(20)

Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY  \
4114         4114  201617123  Total   17123  2016     17     123   
14839       14839  200447095  Total   47095  2004     47      95   
18009       18009  200847149  Total   47149  2008     47     149   
15613       15613   20046063  Total    6063  2004      6      63   
4502         4502  201621103  Total   21103  2016     21     103   
6345         6345  201651141  Total   51141  2016     51     141   
1365         1365  201218073  Total   18073  2012     18      73   
5948         5948  201647125  Total   47125  2016     47     125   
13607       13607  200426055  Total   26055  2004     26      55   
11637       11637  200046111  Total   46111  2000     46     111   
10232       10232  200021103  Total   21103  2000     21     103   
11371       11371  200040063  Total   40063  2000     40      63   
11570       11570  200045067  Total   45067  2000     45      67   
18684       18684  200855107  Total   55107  2008     55     107   
12543       12543   20008091  Total    8091  2000      8      91   
2574         2574  201237105  Total   37105  2012     37     105   
10212       10212  200021065  Total   21065  2000     21      65   
12572       12572   20041001  Total    1001  2004      1       1   
17576       17576  200839083  Total   39083  2008     39      83   
1992         1992  201227093  Total   27093  2012     27      93   

               STNAME                CTYNAME  AA_FEMALE          ...           \
4114         Illinois        Marshall County         31          ...            
14839       Tennessee            Lake County          7          ...            
18009       Tennessee      Rutherford County       3698          ...            
15613      California          Plumas County         78          ...            
4502         Kentucky           Henry County         26          ...            
6345         Virginia         Patrick County         35          ...            
1365          Indiana          Jasper County         81          ...            
5948        Tennessee      Montgomery County       2655          ...            
13607        Michigan  Grand Traverse County        256          ...            
11637    South Dakota         Sanborn County         10          ...            
10232        Kentucky           Henry County         33          ...            
11371        Oklahoma          Hughes County         15          ...            
11570  South Carolina          Marion County         57          ...            
18684       Wisconsin            Rusk County         27          ...            
12543        Colorado           Ouray County          6          ...            
2574   North Carolina             Lee County        342          ...            
10212        Kentucky          Estill County          3          ...            
12572         Alabama         Autauga County        186          ...            
17576            Ohio            Knox County        173          ...            
1992        Minnesota          Meeker County         37          ...            

       WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
4114     0.493101  0.484154     0.504557   0.495443         0.221590   
14839    0.296246  0.381775     0.377051   0.622949         0.197166   
18009    0.415605  0.412727     0.504876   0.495124         0.297479   
15613    0.465620  0.469342     0.498115   0.501885         0.240063   
4502     0.484700  0.464451     0.508807   0.491193         0.260543   
6345     0.470786  0.453785     0.507702   0.492298         0.201529   
1365     0.491701  0.485909     0.502179   0.497821         0.287825   
5948     0.360317  0.365424     0.500736   0.499264         0.296390   
13607    0.493789  0.470544     0.509717   0.490283         0.267879   
11637    0.475336  0.514574     0.483184   0.516816         0.289238   
10232    0.478947  0.474617     0.501865   0.498135         0.278481   
11371    0.3

### Load and merge senate and house election datasets

In [16]:
# load senate and house datasets
senate = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/senate.csv"
house = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/house.csv"

senate_df = pd.read_csv(senate, encoding = "ISO-8859-1")
house_df = pd.read_csv(house, encoding = "ISO-8859-1")

### Clean up senate election data

In [17]:
# check out senate election dataset
print("Senate data dimensions: {}".format(senate_df.shape))
senate_df.sample(5)

Senate data dimensions: (3421, 18)


year         state state_po  state_fips  state_cen  state_ic     office  \
2598  2010        Nevada       NV          32         88        65  US Senate   
501   1982         Texas       TX          48         74        49  US Senate   
1325  1996       Alabama       AL           1         63        41  US Senate   
2547  2010          Iowa       IA          19         42        31  US Senate   
1766  2000  Pennsylvania       PA          42         23        14  US Senate   

       district stage  special      candidate        party  writein   mode  \
2598  statewide   gen    False      Wil Stand  independent    False  total   
501   statewide   gen    False            NaN          NaN     True  total   
1325  statewide   gen    False  Roger Bedford     democrat    False  total   
2547  statewide   gen    False            NaN          NaN     True  total   
1766  statewide   gen    False  Rick Santorum   republican    False  total   

      candidatevotes  totalvotes  unofficial     version  
2598            2119      721404       False  20171011.0  
501              127     3103167       False  20171011.0  
1325          681651     1499393       False  20171011.0  
2547             872     1116063       False  20171011.0  
1766         2481962     4735116       False  20171011.0

In [18]:
# keep only election years data for senate and house elections
election_years = [1998, 2002, 2006, 2010, 2014, 2018]

senate_df = senate_df[senate_df["year"].isin(election_years)]
print("Senate data dimensions: {}".format(senate_df.shape))

Senate data dimensions: (1038, 18)


In [19]:
senate_df.sample(10)

year          state state_po  state_fips  state_cen  state_ic  \
1902  2002  New Hampshire       NH          33         12         4   
2590  2010         Nevada       NV          32         88        65   
2880  2014       Colorado       CO           8         84        62   
3384  2018      Tennessee       TN          47         62        54   
2602  2010  New Hampshire       NH          33         12         4   
3274  2018     California       CA           6         93        71   
3379  2018      Tennessee       TN          47         62        54   
1603  1998   Pennsylvania       PA          42         23        14   
3417  2018        Wyoming       WY          56         83        68   
1882  2002      Minnesota       MN          27         41        33   

         office   district stage  special          candidate        party  \
1902  US Senate  statewide   gen    False        Ken Blevens  libertarian   
2590  US Senate  statewide   gen    False      Jesse Holland  independent   
2880  US Senate  statewide   gen    False         Mark Udall     democrat   
3384  US Senate  statewide   gen    False  Kevin Lee McCants  independent   
2602  US Senate  statewide   gen    False        Chris Booth  independent   
3274  US Senate  statewide   gen    False      Kevin De Leon     democrat   
3379  US Senate  statewide   gen    False   Marsha Blackburn   republican   
1603  US Senate  statewide   gen    False      Arlen Specter   republican   
3417  US Senate  statewide   gen    False      John Barrasso   republican   
1882  US Senate  statewide   gen    False                NaN          NaN   

      writein   mode  candidatevotes  totalvotes  unofficial     version  
1902    False  total            9835      447135       False  20171011.0  
2590    False  total            3175      721404       False  20171011.0  
2880    False  total          944203     2041058       False  20171011.0  
3384    False  total            1927     2243740       False  20190110.0  
2602    False  total            9194      454710       False  20171011.0  
3274    False  total         5093942    11113364       False  20190110.0  
3379    False  total         1227483     2243740       False  20190110.0  
1603    False  total         1814180     2957499       False  20171011.0  
3417    False  total          136210      203420       False  20190110.0  
1882     True  total            1790     2254639       False  20171011.0

In [20]:
# for each senate election in the state, find the winning party

# group each election by year, state, and party
senate_modified = senate_df.groupby(["year", "state", "state_fips", "party"])["party", "candidatevotes"].max().reset_index()

# sort candidate votes by descending to find the candidates with the most votes
senate_modified = senate_modified.sort_values(by=["year", "state", "state_fips", "candidatevotes"], ascending=False)

# get the party with the most votes by election year and state
senate_modified = senate_modified.groupby(["year", "state", "state_fips"]).first().reset_index()

In [21]:
# check out modified senate dataset
senate_modified.head(5)

year       state  state_fips       party  candidatevotes
0  1998     Alabama           1  republican          817973
1  1998      Alaska           2  republican          165227
2  1998     Arizona           4  republican          696577
3  1998    Arkansas           5    democrat          385878
4  1998  California           6    democrat         4410056

In [22]:
# need to recode winning party to republican or democrat
senate_modified["party"].unique()

array(['republican', 'democrat', 'connecticut for lieberman',
       'democratic-farmer-labor', 'independent'], dtype=object)

In [26]:
# code other democratic affiliated parties to democrat
senate_modified = senate_modified.replace(["democrat (not identified on ballot)", "democratic-farmer-labor", 
                                            "democratic-nonpartisan league"], "democrat")

# code indepdent affiliated parties to independent
senate_modified = senate_modified.replace("independent for Maine", "independent")
senate_modified = senate_modified.replace("connecticut for lieberman", "independent")

# filter out independent party elections
senate_modified = senate_modified[senate_modified.party != "independent"]

In [27]:
# verify only republican and democrat winning elections remain
senate_modified["party"].unique()

array(['republican', 'democrat'], dtype=object)

In [28]:
# create one hot encoded column for winning party
senate_binary = pd.get_dummies(senate_modified["party"], drop_first = True)
senate_modified["senate_winning_party"] = senate_binary

In [30]:
print("Modified senate dataset dimensions: {}".format(senate_modified.shape))
senate_modified.head(5)

Modified senate dataset dimensions: (200, 6)


year       state  state_fips       party  candidatevotes  \
0  1998     Alabama           1  republican          817973   
1  1998      Alaska           2  republican          165227   
2  1998     Arizona           4  republican          696577   
3  1998    Arkansas           5    democrat          385878   
4  1998  California           6    democrat         4410056   

   senate_winning_party  
0                     1  
1                     1  
2                     1  
3                     0  
4                     0

In [31]:
# see how many nulls there are within dataframe
senate_modified.isnull().sum()

year                    0
state                   0
state_fips              0
party                   0
candidatevotes          0
senate_winning_party    0
dtype: int64

In [32]:
#This shows how many states had senate elections
senate_modified['year'].value_counts()

2010    36
2014    34
2002    34
1998    34
2018    31
2006    31
Name: year, dtype: int64

### Clean up house election data

In [33]:
# checkout house election dataset
print("House data dimensions: {}".format(house_df.shape))
house_df.head(10)

House data dimensions: (29636, 19)


year    state state_po  state_fips  state_cen  state_ic    office  \
0  1976  Alabama       AL           1         63        41  US House   
1  1976  Alabama       AL           1         63        41  US House   
2  1976  Alabama       AL           1         63        41  US House   
3  1976  Alabama       AL           1         63        41  US House   
4  1976  Alabama       AL           1         63        41  US House   
5  1976  Alabama       AL           1         63        41  US House   
6  1976  Alabama       AL           1         63        41  US House   
7  1976  Alabama       AL           1         63        41  US House   
8  1976  Alabama       AL           1         63        41  US House   
9  1976  Alabama       AL           1         63        41  US House   

   district stage runoff  special                      candidate        party  \
0         1   gen  False    False                 Bill Davenport     democrat   
1         1   gen  False    False                   Jack Edwards   republican   
2         1   gen  False    False                            NaN          NaN   
3         2   gen  False    False               J. Carole Keahey     democrat   
4         2   gen  False    False                            NaN          NaN   
5         2   gen  False    False  William L. ""Bill"" Dickinson   republican   
6         3   gen  False    False                   Bill Nichols     democrat   
7         3   gen  False    False                            NaN          NaN   
8         3   gen  False    False                 Ogburn Gardner  prohibition   
9         4   gen  False    False                     Tom Bevill     democrat   

   writein   mode  candidatevotes  totalvotes  unofficial   version  
0    False  total           58906      157170       False  20200424  
1    False  total           98257      157170       False  20200424  
2     True  total               7      157170       False  20200424  
3    False  total           66288      156362       False  20200424  
4     True  total               5      156362       False  20200424  
5    False  total           90069      156362       False  20200424  
6    False  total          106935      108048       False  20200424  
7     True  total               2      108048       False  20200424  
8    False  total            1111      108048       False  20200424  
9    False  total          141490      176022       False  20200424

In [34]:
# filter for only election years data for senate and house elections
house_df = house_df[house_df["year"].isin(election_years)]
print("House data dimensions: {}".format(house_df.shape))

House data dimensions: (8396, 19)


In [35]:
# house election dataset seems to be structured similarly to senate election dataset
house_modified = house_df.groupby(["year", "state", "state_fips", "party"])["party", "candidatevotes"].max().reset_index()

# sort candidate votes by descending to find the candidates with the most votes
house_modified = house_modified.sort_values(by=["year", "state", "state_fips", "candidatevotes"], ascending=False)

# get the party with the most votes by election year and state
house_modified = house_modified.groupby(["year", "state", "state_fips"]).first().reset_index()

In [36]:
house_modified.head(5)

year       state  state_fips       party  candidatevotes
0  1998     Alabama           1  republican          154761
1  1998      Alaska           2  republican          139676
2  1998     Arizona           4  republican          137618
3  1998    Arkansas           5  republican          154780
4  1998  California           6    democrat          158446

In [37]:
# need to recode winning party to republican or democrat
house_modified["party"].unique()

array(['republican', 'democrat', 'independent', 'democratic-farmer-labor'], dtype=object)

In [38]:
# code other democratic affiliated parties to democrat
house_modified = house_modified.replace("democratic-farmer-labor", "democrat")

# filter out independent party elections
house_modified = house_modified[house_modified.party != "independent"]

In [39]:
# verify only republican and democrat winning elections remain
house_modified["party"].unique()

array(['republican', 'democrat'], dtype=object)

In [40]:
# create one hot encoded column for winning party
house_binary = pd.get_dummies(house_modified["party"], drop_first = True)
house_modified["house_winning_party"] = house_binary

In [41]:
print("Modified house dataset dimensions: {}".format(house_modified.shape))
house_modified.head(10)

Modified house dataset dimensions: (298, 6)


year        state  state_fips       party  candidatevotes  \
0  1998      Alabama           1  republican          154761   
1  1998       Alaska           2  republican          139676   
2  1998      Arizona           4  republican          137618   
3  1998     Arkansas           5  republican          154780   
4  1998   California           6    democrat          158446   
5  1998     Colorado           8  republican          156501   
6  1998  Connecticut           9    democrat          109726   
7  1998     Delaware          10  republican          119811   
8  1998      Florida          12  republican          140525   
9  1998      Georgia          13  republican          164966   

   house_winning_party  
0                    1  
1                    1  
2                    1  
3                    1  
4                    0  
5                    1  
6                    0  
7                    1  
8                    1  
9                    1

In [42]:
# see how many nulls there are within dataframe
house_modified.isnull().sum()

year                   0
state                  0
state_fips             0
party                  0
candidatevotes         0
house_winning_party    0
dtype: int64

In [43]:
#This shows how many states had senate elections
house_modified['year'].value_counts()

2018    50
2014    50
2010    50
2006    50
2002    49
1998    49
Name: year, dtype: int64

### Merge senate and house election dataset 

In [44]:
# sort candidate votes by descending to find the candidates with the most votes by district
house_senate_merged = pd.merge(house_modified[["year", "state", "state_fips", "house_winning_party"]],
                               senate_modified[["year", "state", "state_fips", "senate_winning_party"]],
                               how="left", left_on=["year", "state_fips"], right_on=["year", "state_fips"])

In [45]:
house_senate_merged.head()

year     state_x  state_fips  house_winning_party     state_y  \
0  1998     Alabama           1                    1     Alabama   
1  1998      Alaska           2                    1      Alaska   
2  1998     Arizona           4                    1     Arizona   
3  1998    Arkansas           5                    1    Arkansas   
4  1998  California           6                    0  California   

   senate_winning_party  
0                   1.0  
1                   1.0  
2                   1.0  
3                   0.0  
4                   0.0

In [46]:
# verify data types
house_senate_merged.dtypes

year                      int64
state_x                  object
state_fips                int64
house_winning_party       uint8
state_y                  object
senate_winning_party    float64
dtype: object

In [47]:
# see how many nulls there are within dataframe
house_senate_merged.isnull().sum()

year                     0
state_x                  0
state_fips               0
house_winning_party      0
state_y                 99
senate_winning_party    99
dtype: int64

In [48]:
house_senate_merged["senate_winning_party"].fillna(-1, inplace = True)
house_senate_merged["house_winning_party"].fillna(-1, inplace = True)

In [49]:
# see how many nulls there are within dataframe
house_senate_merged.isnull().sum()

year                     0
state_x                  0
state_fips               0
house_winning_party      0
state_y                 99
senate_winning_party     0
dtype: int64

In [52]:
#replace values to the election years to enable the merge
replace_values = { 1998 : 2000, 2002 : 2004, 2006 : 2008 , 2010 : 2012, 2014 : 2016, 2018 : 2020 }     
house_senate_merged = house_senate_merged.replace({"year": replace_values})  
house_senate_merged.sample(10)

year        state_x  state_fips  house_winning_party        state_y  \
134  2008         Oregon          41                    0            NaN   
277  2020     New Jersey          34                    0     New Jersey   
117  2008       Maryland          24                    1       Maryland   
170  2012      Minnesota          27                    1            NaN   
69   2004  Massachusetts          25                    0  Massachusetts   
82   2004   North Dakota          38                    0            NaN   
135  2008   Pennsylvania          42                    0   Pennsylvania   
169  2012       Michigan          26                    1            NaN   
65   2004       Kentucky          21                    1       Kentucky   
89   2004   South Dakota          46                    1   South Dakota   

     senate_winning_party  
134                  -1.0  
277                   0.0  
117                   0.0  
170                  -1.0  
69                    0.0  
82                   -1.0  
135                   0.0  
169                  -1.0  
65                    1.0  
89                    0.0

In [53]:
election_years = [2020]

F20_house_senate = house_senate_merged[house_senate_merged["year"].isin(election_years)]
print("2020 data dimensions: {}".format(F20_house_senate.shape))

2020 data dimensions: (50, 6)


In [54]:
election_years = [2000, 2004, 2008, 2012, 2016]

house_senate_merged = house_senate_merged[house_senate_merged["year"].isin(election_years)]
print("senate house data dimensions: {}".format(house_senate_merged.shape))

senate house data dimensions: (248, 6)


### Merge senate and house election dataset to Ahana's total dataset

In [55]:
# verify dataset dimensions before merge
print("Ahana's dataset dimensions: {}".format(merged_df2.shape))

Ahana's dataset dimensions: (15569, 65)


In [56]:
# merge dataframes on election year and state fips code
combined_df = pd.merge(merged_df2, house_senate_merged, how="left", 
                       left_on=["YEAR", "STATE"], right_on=["year", "state_fips"])

In [57]:
print("Combined dataset dimensions: {}".format(combined_df.shape))
combined_df.head()

Combined dataset dimensions: (15569, 71)


Unnamed: 0        ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY   STNAME  \
0           0  20121001  Total    1001  2012      1       1  Alabama   
1           1  20121003  Total    1003  2012      1       3  Alabama   
2           2  20121005  Total    1005  2012      1       5  Alabama   
3           3  20121007  Total    1007  2012      1       7  Alabama   
4           4  20121009  Total    1009  2012      1       9  Alabama   

          CTYNAME  AA_FEMALE          ...           TOT_POP_40to59%  \
0  Autauga County        296          ...                  0.282427   
1  Baldwin County        833          ...                  0.282293   
2  Barbour County         59          ...                  0.278741   
3     Bibb County         10          ...                  0.280955   
4   Blount County         82          ...                  0.276594   

   TOT_POP_Above60%  Unemployment_Rate  winning_party_binary    year  state_x  \
0          0.171659                6.9                     1  2012.0  Alabama   
1          0.234071                7.5                     1  2012.0  Alabama   
2          0.208233               11.5                     0  2012.0  Alabama   
3          0.187128                8.5                     1  2012.0  Alabama   
4          0.210691                6.9                     1  2012.0  Alabama   

   state_fips  house_winning_party  state_y  senate_winning_party  
0         1.0                  1.0  Alabama                   1.0  
1         1.0                  1.0  Alabama                   1.0  
2         1.0                  1.0  Alabama                   1.0  
3         1.0                  1.0  Alabama                   1.0  
4         1.0                  1.0  Alabama                   1.0  

[5 rows x 71 columns]

In [58]:
combined_df.columns

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       '

In [59]:
# drop unnecessary columns
dropCombinedCols = ["SUMLEV", "AGEGRP", "FIPS_x", "year", "state_x", "state_fips", "state_y"]
combined_df = combined_df.drop(dropCombinedCols,axis=1)

In [60]:
combined_df.columns

Index(['Unnamed: 0', 'ID', 'YEAR', 'STATE', 'COUNTY', 'STNAME', 'CTYNAME',
       'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
       'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE',
       'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19',
       'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 'TOT_POP_20to39',
       'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59',
       'TOT_MALE_40to59', 'TOT_FEMALE_40to59', 'TOT_POP_Above60',
       'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'republican_candidate',
       'democrat_candidate', 'republican_votes', 'democrat_votes',
       'county_totalvotes', 'winning candidate', 'winning party',
       'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%',
       'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 'NA_FEMALE%',
       'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%',
       'TOT_POP_LESS19%', 'TOT_POP_20to39%', 'TOT_POP_40to59%',
       'TOT_POP_Abo

In [61]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15569 entries, 0 to 15568
Data columns (total 64 columns):
Unnamed: 0              15569 non-null int64
ID                      15569 non-null int64
YEAR                    15569 non-null int64
STATE                   15569 non-null int64
COUNTY                  15569 non-null int64
STNAME                  15569 non-null object
CTYNAME                 15569 non-null object
AA_FEMALE               15569 non-null int64
AA_MALE                 15569 non-null int64
BA_FEMALE               15569 non-null int64
BA_MALE                 15569 non-null int64
H_FEMALE                15569 non-null int64
H_MALE                  15569 non-null int64
IA_FEMALE               15569 non-null int64
IA_MALE                 15569 non-null int64
NA_FEMALE               15569 non-null int64
NA_MALE                 15569 non-null int64
TOT_FEMALE              15569 non-null int64
TOT_MALE                15569 non-null int64
TOT_POP                 15569 non-

In [62]:
# rename columns
renameCombinedCols = ['ID','YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
       'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
       'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE',
       'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19',
       'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 'TOT_POP_20to39',
       'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59',
       'TOT_MALE_40to59', 'TOT_FEMALE_40to59', 'TOT_POP_Above60',
       'TOT_MALE_Above60', 'TOT_FEMALE_Above60',
       'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES',
       'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
       'WINNING_PARTY', 'AVG_WAGE_SALARY', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       'TOT_POP_40to59%', 'TOT_POP_Above60%', 'UNEMPLOYMENT_RATE',
       'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY']

len(renameCombinedCols)
combined_df.columns = renameCombinedCols
combined_df.head()



ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS    STATE          COUNTY  \
0   0   20121001  2012           1            1  Alabama  Autauga County   
1   1   20121003  2012           1            3  Alabama  Baldwin County   
2   2   20121005  2012           1            5  Alabama  Barbour County   
3   3   20121007  2012           1            7  Alabama     Bibb County   
4   4   20121009  2012           1            9  Alabama   Blount County   

   AA_FEMALE  AA_MALE  BA_FEMALE          ...            TOT_FEMALE%  \
0        296      201       5152          ...               0.512912   
1        833      551       8913          ...               0.510581   
2         59       59       6111          ...               0.469301   
3         10       12       1972          ...               0.463799   
4         82       48        390          ...               0.505255   

   TOT_MALE%  TOT_POP_LESS19%  TOT_POP_20to39%  TOT_POP_40to59%  \
0   0.487088         0.293056         0.252858         0.282427   
1   0.489419         0.253453         0.230183         0.282293   
2   0.530699         0.242225         0.270801         0.278741   
3   0.536201         0.248820         0.283097         0.280955   
4   0.494745         0.271033         0.241682         0.276594   

   TOT_POP_Above60%  UNEMPLOYMENT_RATE  WINNING_PARTY_BINARY  \
0          0.171659                6.9                     1   
1          0.234071                7.5                     1   
2          0.208233               11.5                     0   
3          0.187128                8.5                     1   
4          0.210691                6.9                     1   

   HOUSE_WINNING_BINARY  SENATE_WINNING_BINARY  
0                   1.0                    1.0  
1                   1.0                    1.0  
2                   1.0                    1.0  
3                   1.0                    1.0  
4                   1.0                    1.0  

[5 rows x 64 columns]

In [63]:
# reorder columns
reorderCombinedCols = ['ID','YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
                            'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES', 'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
                            'WINNING_PARTY', 'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY','UNEMPLOYMENT_RATE',
                            'AVG_WAGE_SALARY', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
                            'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 
                            'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 
                            'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59', 'TOT_MALE_40to59', 
                            'TOT_FEMALE_40to59', 'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'AA_FEMALE%', 
                            'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 
                            'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 
                            'TOT_POP_20to39%', 'TOT_POP_40to59%', 'TOT_POP_Above60%']

combined_df = combined_df[['ID','YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
                            'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES', 'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
                            'WINNING_PARTY', 'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY','UNEMPLOYMENT_RATE',
                            'AVG_WAGE_SALARY', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
                            'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 
                            'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 
                            'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59', 'TOT_MALE_40to59', 
                            'TOT_FEMALE_40to59', 'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'AA_FEMALE%', 
                            'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 
                            'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 
                            'TOT_POP_20to39%', 'TOT_POP_40to59%', 'TOT_POP_Above60%']]

In [64]:
# check out final combined dataset
combined_df.head()

ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS    STATE          COUNTY  \
0   0   20121001  2012           1            1  Alabama  Autauga County   
1   1   20121003  2012           1            3  Alabama  Baldwin County   
2   2   20121005  2012           1            5  Alabama  Barbour County   
3   3   20121007  2012           1            7  Alabama     Bibb County   
4   4   20121009  2012           1            9  Alabama   Blount County   

  REP_CANDIDATE DEM_CANDIDATE  REP_VOTES        ...         NA_FEMALE%  \
0   Mitt Romney  Barack Obama    17379.0        ...           0.000310   
1   Mitt Romney  Barack Obama    66016.0        ...           0.000300   
2   Mitt Romney  Barack Obama     5550.0        ...           0.000622   
3   Mitt Romney  Barack Obama     6132.0        ...           0.000219   
4   Mitt Romney  Barack Obama    20757.0        ...           0.000436   

   NA_MALE% WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
0  0.000511   0.402747  0.390054     0.512912   0.487088         0.293056   
1  0.000399   0.446461  0.429892     0.510581   0.489419         0.253453   
2  0.001061   0.237175  0.271277     0.469301   0.530699         0.242225   
3  0.001137   0.371240  0.394981     0.463799   0.536201         0.248820   
4  0.000627   0.487947  0.477838     0.505255   0.494745         0.271033   

   TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  
0         0.252858         0.282427          0.171659  
1         0.230183         0.282293          0.234071  
2         0.270801         0.278741          0.208233  
3         0.283097         0.280955          0.187128  
4         0.241682         0.276594          0.210691  

[5 rows x 64 columns]

In [65]:
# see how many nulls there are within dataframe
combined_df.isnull().sum()

ID                        0
YEAR_FIPS                 0
YEAR                      0
STATE_FIPS                0
COUNTY_FIPS               0
STATE                     0
COUNTY                    0
REP_CANDIDATE             0
DEM_CANDIDATE             0
REP_VOTES                 0
DEM_VOTES                 0
COUNTY_TOTALVOTES         0
WINNING_CANDIDATE         0
WINNING_PARTY             0
WINNING_PARTY_BINARY      0
HOUSE_WINNING_BINARY     33
SENATE_WINNING_BINARY    33
UNEMPLOYMENT_RATE         0
AVG_WAGE_SALARY           0
AA_FEMALE                 0
AA_MALE                   0
BA_FEMALE                 0
BA_MALE                   0
H_FEMALE                  0
H_MALE                    0
IA_FEMALE                 0
IA_MALE                   0
NA_FEMALE                 0
NA_MALE                   0
TOT_FEMALE                0
                         ..
TOT_POP_LESS19            0
TOT_MALE_LESS19           0
TOT_FEMALE_LESS19         0
TOT_POP_20to39            0
TOT_MALE_20to39     

In [66]:
combined_df["HOUSE_WINNING_BINARY"].fillna(-1, inplace = True)
combined_df["SENATE_WINNING_BINARY"].fillna(-1, inplace = True)

In [67]:
#converting House winning binary and State Winnng binary to integers
convert_dict = {'HOUSE_WINNING_BINARY': int, 
                'SENATE_WINNING_BINARY': int
               } 
  
combined_df = combined_df.astype(convert_dict)


In [68]:
#Adding Margin of victory for plotting - Ranging from Red to Blue
combined_df['MARGIN_VICTORY']=(combined_df['DEM_VOTES']/combined_df['COUNTY_TOTALVOTES'])-(combined_df['REP_VOTES']/combined_df['COUNTY_TOTALVOTES'])


### Save one hot encoded dataframe to csv file

In [69]:
# output one hot encoded dataframe with null values to csv file
combined_df.to_csv("combined_jun27.csv")

## Merge senate and house election dataset to F20 Dataset

In [105]:
# verify dataset dimensions before merge
print("F20 demographic: {}".format(F20_df.shape))

F20 demographic: (3142, 65)


In [106]:
# merge dataframes on election year and state fips code
combined_df_20 = pd.merge(F20_df, F20_house_senate, how="left", 
                       left_on=["YEAR", "STATE"], right_on=["year", "state_fips"])

In [107]:
print("Combined F20 dataset dimensions: {}".format(combined_df_20.shape))
combined_df_20.head()

Combined F20 dataset dimensions: (3142, 71)


Unnamed: 0        ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY   STNAME  \
0         632  20201001  Total    1001  2020      1       1  Alabama   
1         633  20201003  Total    1003  2020      1       3  Alabama   
2         634  20201005  Total    1005  2020      1       5  Alabama   
3         635  20201007  Total    1007  2020      1       7  Alabama   
4         636  20201009  Total    1009  2020      1       9  Alabama   

          CTYNAME  AA_FEMALE          ...           TOT_POP_40to59%  \
0  Autauga County        380          ...                  0.275535   
1  Baldwin County       1575          ...                  0.262928   
2  Barbour County         61          ...                  0.255295   
3     Bibb County         34          ...                  0.274687   
4   Blount County        103          ...                  0.264834   

   TOT_POP_Above60%  Unemployment_Rate  winning_party_binary    year  state_x  \
0          0.213845                NaN                     0  2020.0  Alabama   
1          0.276284                NaN                     0  2020.0  Alabama   
2          0.255938                NaN                     0  2020.0  Alabama   
3          0.225580                NaN                     0  2020.0  Alabama   
4          0.246041                NaN                     0  2020.0  Alabama   

   state_fips  house_winning_party  state_y  senate_winning_party  
0         1.0                  1.0      NaN                  -1.0  
1         1.0                  1.0      NaN                  -1.0  
2         1.0                  1.0      NaN                  -1.0  
3         1.0                  1.0      NaN                  -1.0  
4         1.0                  1.0      NaN                  -1.0  

[5 rows x 71 columns]

In [108]:
combined_df_20.columns

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       '

In [109]:
combined_df_20 = combined_df_20.drop(dropCombinedCols,axis=1)

In [110]:
combined_df_20.columns = renameCombinedCols
combined_df_20.head()

ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS    STATE          COUNTY  \
0  632   20201001  2020           1            1  Alabama  Autauga County   
1  633   20201003  2020           1            3  Alabama  Baldwin County   
2  634   20201005  2020           1            5  Alabama  Barbour County   
3  635   20201007  2020           1            7  Alabama     Bibb County   
4  636   20201009  2020           1            9  Alabama   Blount County   

   AA_FEMALE  AA_MALE  BA_FEMALE          ...            TOT_FEMALE%  \
0        380      301       5762          ...               0.514487   
1       1575      933      10092          ...               0.515384   
2         61       52       5672          ...               0.472168   
3         34       19       1816          ...               0.467813   
4        103       82        443          ...               0.507261   

   TOT_MALE%  TOT_POP_LESS19%  TOT_POP_20to39%  TOT_POP_40to59%  \
0   0.485513         0.259420         0.251201         0.275535   
1   0.484616         0.236994         0.223794         0.262928   
2   0.527832         0.228769         0.259998         0.255295   
3   0.532188         0.225536         0.274196         0.274687   
4   0.492739         0.253544         0.235581         0.264834   

   TOT_POP_Above60%  UNEMPLOYMENT_RATE  WINNING_PARTY_BINARY  \
0          0.213845                NaN                     0   
1          0.276284                NaN                     0   
2          0.255938                NaN                     0   
3          0.225580                NaN                     0   
4          0.246041                NaN                     0   

   HOUSE_WINNING_BINARY  SENATE_WINNING_BINARY  
0                   1.0                   -1.0  
1                   1.0                   -1.0  
2                   1.0                   -1.0  
3                   1.0                   -1.0  
4                   1.0                   -1.0  

[5 rows x 64 columns]

In [111]:
# reorder columns
reorderCombinedCols = ['ID','YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
                            'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES', 'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
                            'WINNING_PARTY', 'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY','UNEMPLOYMENT_RATE',
                            'AVG_WAGE_SALARY', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
                            'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 
                            'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 
                            'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59', 'TOT_MALE_40to59', 
                            'TOT_FEMALE_40to59', 'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'AA_FEMALE%', 
                            'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 
                            'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 
                            'TOT_POP_20to39%', 'TOT_POP_40to59%', 'TOT_POP_Above60%']

combined_df_20 = combined_df_20[['ID','YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
                            'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES', 'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
                            'WINNING_PARTY', 'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY','UNEMPLOYMENT_RATE',
                            'AVG_WAGE_SALARY', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
                            'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 
                            'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 
                            'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59', 'TOT_MALE_40to59', 
                            'TOT_FEMALE_40to59', 'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'AA_FEMALE%', 
                            'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 
                            'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 
                            'TOT_POP_20to39%', 'TOT_POP_40to59%', 'TOT_POP_Above60%']]

In [112]:
# see how many nulls there are within dataframe
combined_df_20.isnull().sum()

ID                          0
YEAR_FIPS                   0
YEAR                        0
STATE_FIPS                  0
COUNTY_FIPS                 0
STATE                       0
COUNTY                      0
REP_CANDIDATE            3142
DEM_CANDIDATE            3142
REP_VOTES                3142
DEM_VOTES                3142
COUNTY_TOTALVOTES        3142
WINNING_CANDIDATE        3142
WINNING_PARTY            3142
WINNING_PARTY_BINARY        0
HOUSE_WINNING_BINARY        1
SENATE_WINNING_BINARY       1
UNEMPLOYMENT_RATE        3142
AVG_WAGE_SALARY          3142
AA_FEMALE                   0
AA_MALE                     0
BA_FEMALE                   0
BA_MALE                     0
H_FEMALE                    0
H_MALE                      0
IA_FEMALE                   0
IA_MALE                     0
NA_FEMALE                   0
NA_MALE                     0
TOT_FEMALE                  0
                         ... 
TOT_POP_LESS19              0
TOT_MALE_LESS19             0
TOT_FEMALE

In [113]:
#copying 2016 unemployment data and average wage salary into 2020
election_years = [2016]

F16_df = merged_df[merged_df["YEAR"].isin(election_years)]
print("2016 data dimensions: {}".format(F16_df.shape))

2016 data dimensions: (3142, 65)


In [114]:
F16_df.head(10)

Unnamed: 0        ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY   STNAME  \
316         316  20161001  Total    1001  2016      1       1  Alabama   
317         317  20161003  Total    1003  2016      1       3  Alabama   
318         318  20161005  Total    1005  2016      1       5  Alabama   
319         319  20161007  Total    1007  2016      1       7  Alabama   
320         320  20161009  Total    1009  2016      1       9  Alabama   
321         321  20161011  Total    1011  2016      1      11  Alabama   
322         322  20161013  Total    1013  2016      1      13  Alabama   
323         323  20161015  Total    1015  2016      1      15  Alabama   
324         324  20161017  Total    1017  2016      1      17  Alabama   
325         325  20161019  Total    1019  2016      1      19  Alabama   

             CTYNAME  AA_FEMALE          ...           WA_FEMALE%  WA_MALE%  \
316   Autauga County        337          ...             0.398006  0.384230   
317   Baldwin County       1154          ...             0.445843  0.426208   
318   Barbour County         55          ...             0.233052  0.266799   
319      Bibb County         23          ...             0.373275  0.394348   
320    Blount County         93          ...             0.487161  0.474314   
321   Bullock County         10          ...             0.100422  0.161275   
322    Butler County        106          ...             0.279703  0.253453   
323   Calhoun County        626          ...             0.389844  0.368490   
324  Chambers County        153          ...             0.297763  0.281313   
325  Cherokee County         47          ...             0.468960  0.463787   

     TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  TOT_POP_20to39%  \
316     0.511918   0.488082         0.272432         0.250747   
317     0.512247   0.487753         0.244050         0.228220   
318     0.467673   0.532327         0.231706         0.269639   
319     0.461470   0.538530         0.229759         0.276696   
320     0.506806   0.493194         0.258097         0.235566   
321     0.453633   0.546367         0.229161         0.281294   
322     0.535516   0.464484         0.252913         0.239542   
323     0.519169   0.480831         0.247321         0.261761   
324     0.520718   0.479282         0.235727         0.235786   
325     0.502162   0.497838         0.225041         0.206625   

     TOT_POP_40to59%  TOT_POP_Above60%  Unemployment_Rate  \
316         0.286373          0.190448                5.1   
317         0.276195          0.251535                5.3   
318         0.269340          0.229315                8.3   
319         0.287698          0.205847                6.4   
320         0.274908          0.231429                5.4   
321         0.280075          0.209470                6.8   
322         0.256648          0.250897                6.9   
323         0.266624          0.224294                6.5   
324         0.278458          0.250029                5.4   
325         0.286928          0.281407                5.0   

     winning_party_binary  
316                     1  
317                     1  
318                     1  
319                     1  
320                     1  
321                     0  
322                     1  
323                     1  
324                     1  
325                     1  

[10 rows x 65 columns]

In [115]:
combined_df_20.head(10)

ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS    STATE           COUNTY  \
0  632   20201001  2020           1            1  Alabama   Autauga County   
1  633   20201003  2020           1            3  Alabama   Baldwin County   
2  634   20201005  2020           1            5  Alabama   Barbour County   
3  635   20201007  2020           1            7  Alabama      Bibb County   
4  636   20201009  2020           1            9  Alabama    Blount County   
5  637   20201011  2020           1           11  Alabama   Bullock County   
6  638   20201013  2020           1           13  Alabama    Butler County   
7  639   20201015  2020           1           15  Alabama   Calhoun County   
8  640   20201017  2020           1           17  Alabama  Chambers County   
9  641   20201019  2020           1           19  Alabama  Cherokee County   

  REP_CANDIDATE DEM_CANDIDATE  REP_VOTES        ...         NA_FEMALE%  \
0           NaN           NaN        NaN        ...           0.000522   
1           NaN           NaN        NaN        ...           0.000335   
2           NaN           NaN        NaN        ...           0.000643   
3           NaN           NaN        NaN        ...           0.000179   
4           NaN           NaN        NaN        ...           0.000450   
5           NaN           NaN        NaN        ...           0.003551   
6           NaN           NaN        NaN        ...           0.000254   
7           NaN           NaN        NaN        ...           0.000560   
8           NaN           NaN        NaN        ...           0.000149   
9           NaN           NaN        NaN        ...           0.000115   

   NA_MALE% WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
0  0.000594   0.392097  0.375155     0.514487   0.485513         0.259420   
1  0.000335   0.448716  0.424136     0.515384   0.484616         0.236994   
2  0.001206   0.233069  0.257626     0.472168   0.527832         0.228769   
3  0.000982   0.377188  0.391161     0.467813   0.532188         0.225536   
4  0.000761   0.486670  0.472113     0.507261   0.492739         0.253544   
5  0.003946   0.103176  0.159302     0.454823   0.545177         0.229631   
6  0.000254   0.273577  0.248933     0.534299   0.465701         0.244512   
7  0.000543   0.387375  0.365410     0.519466   0.480534         0.243076   
8  0.000297   0.294244  0.276394     0.521255   0.478745         0.228499   
9  0.000192   0.469960  0.460318     0.504610   0.495390         0.212623   

   TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  
0         0.251201         0.275535          0.213845  
1         0.223794         0.262928          0.276284  
2         0.259998         0.255295          0.255938  
3         0.274196         0.274687          0.225580  
4         0.235581         0.264834          0.246041  
5         0.280233         0.259420          0.230716  
6         0.234909         0.244360          0.276220  
7         0.256640         0.254224          0.246060  
8         0.241380         0.262948          0.267172  
9         0.205747         0.271858          0.309773  

[10 rows x 64 columns]

In [116]:
F16_df.columns

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       '

In [118]:
# output one hot encoded dataframe with null values to csv file
combined_df_20.to_csv("F20_jun27.csv")

In [119]:
# output one hot encoded dataframe with null values to csv file
F16_df.to_csv("F16_jun27.csv")

In [120]:
# counties that flipped between the two parties in the last 20 years.
state_county_map = {}
county_groupby = combined_df.groupby("COUNTY_FIPS")
count = 0
for name, group in county_groupby:
    winning_party = None
    prev_winning_party = None
    for index, row in group.iterrows():
        winning_party = row['WINNING_PARTY']
        
        if prev_winning_party == None:
            prev_winning_party = winning_party
        
        if prev_winning_party != winning_party:
            #break and move to next county...
            #print("training set add: raw: state={}, county={}".format(raw['STATE'], raw['COUNTY']))
            count +=1
            state_county_map[str(row['STATE_FIPS']) + '_' + str(row['COUNTY_FIPS'])] = 1
            break

print(count)

151


In [121]:
#Creating the train Data
#add new column to help us identify county/state combo that should be used in training. 
count_0 = 0
count_1 = 0
combined_df["TRAIN"] = ""
for index, row in combined_df.iterrows():
    state_fips = row['STATE_FIPS']
    county_fips = row['COUNTY_FIPS']
    
    key = str(state_fips) + '_' + str(county_fips)
    
    if state_county_map.get(key) == None:
        combined_df['TRAIN'][index] = 0
        count_0 += 1
    else:
        combined_df['TRAIN'][index] = 1
        count_1 += 1
    
print(count_0)
print(count_1)

C:\Users\nerella\Anaconda3\lib\site-packages\ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  del sys.path[0]
C:\Users\nerella\Anaconda3\lib\site-packages\ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


14815
754


In [122]:
combined_df.loc[combined_df['TRAIN'] == 1]

ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS        STATE  \
68        68   20122016  2012           2           16       Alaska   
69        69   20122020  2012           2           20       Alaska   
70        96   20124001  2012           4            1      Arizona   
75       101   20124011  2012           4           11      Arizona   
80       106   20124019  2012           4           19      Arizona   
82       108   20124023  2012           4           23      Arizona   
87       113   20125005  2012           5            5     Arkansas   
93       119   20125017  2012           5           17     Arkansas   
105      131   20125041  2012           5           41     Arkansas   
108      134   20125047  2012           5           47     Arkansas   
116      142   20125063  2012           5           63     Arkansas   
117      143   20125065  2012           5           65     Arkansas   
119      145   20125069  2012           5           69     Arkansas   
121      147   20125073  2012           5           73     Arkansas   
123      149   20125077  2012           5           77     Arkansas   
127      153   20125085  2012           5           85     Arkansas   
128      154   20125087  2012           5           87     Arkansas   
130      156   20125091  2012           5           91     Arkansas   
135      161   20125101  2012           5          101     Arkansas   
137      163   20125105  2012           5          105     Arkansas   
138      164   20125107  2012           5          107     Arkansas   
141      167   20125113  2012           5          113     Arkansas   
146      172   20125123  2012           5          123     Arkansas   
150      176   20125131  2012           5          131     Arkansas   
158      184   20125147  2012           5          147     Arkansas   
161      187   20126003  2012           6            3   California   
166      192   20126013  2012           6           13   California   
172      198   20126025  2012           6           25   California   
176      202   20126033  2012           6           33   California   
177      203   20126035  2012           6           35   California   
...      ...        ...   ...         ...          ...          ...   
15445  18731   20086013  2008           6           13   California   
15451  18737   20086025  2008           6           25   California   
15455  18741   20086033  2008           6           33   California   
15456  18742   20086035  2008           6           35   California   
15457  18743   20086037  2008           6           37   California   
15461  18747   20086045  2008           6           45   California   
15464  18750   20086051  2008           6           51   California   
15465  18751   20086053  2008           6           53   California   
15466  18752   20086055  2008           6           55   California   
15467  18753   20086057  2008           6           57   California   
15472  18758   20086067  2008           6           67   California   
15474  18760   20086071  2008           6           71   California   
15476  18762   20086075  2008           6           75   California   
15478  18764   20086079  2008           6           79   California   
15479  18765   20086081  2008           6           81   California   
15480  18766   20086083  2008           6           83   California   
15486  18772   20086095  2008           6           95   California   
15487  18773   20086097  2008           6           97   California   
15488  18774   20086099  2008           6           99   California   
15494  18780   20086111  2008           6          111   California   
15504  18790   20088014  2008           8           14     Colorado   
15505  18791   20088015  2008           8           15     Colorado   
15508  18794   20088021  2008           8           21     Colorado   
15513  18799   20088031  2008           8           31     Colorado   
15527  18813   20088059  2008     

In [123]:
# output one hot encoded dataframe with null values to csv file
combined_df.to_csv("combined_jun27_train.csv")